# sup45-ts_readthrough
Analysis pipeline, codes, and processed data for Mangkalaphiban et al.<br/>
DOI: [pending]
Preprint: https://doi.org/10.1101/2020.12.15.422930<br/>

Raw sequencing data generated in this study are deposited deposited and available at Gene Expression Omnibus (GEO) under accession number GSE162780.<br/>

## Analysis pipeline
1. Sequence alignment
	* Input:
		* Raw sequencing data used in this study and their associated SRR numbers are listed in *Raw_data_SRR.csv*
	* Transcriptome used for sequence alignment is available at https://github.com/Jacobson-Lab/yeast_transcriptome_v5
	* Output: 
		* Gene and transcript abundance files generated by RSEM are in the folder _**RSEM_results**_
		* bam files
2. Generate data table containing read information using riboWaltz
	* riboWaltz: https://github.com/LabTranslationalArchitectomics/riboWaltz
	* Input:
		* bam files from step 1 
		* annotation file (***folder**/genedf_riboWaltz_v5_CDS_corrected.txt*)
	* Output:
		* ***folder**/(dataset)_reads_psite_list.Rdata* contains the following reads information: length of read, 5' & 3' ends + position of the read's P-site relative to the annotation provided in the annotation file. Distance from read's P-site to start and stop codons are calculated. The mRNA region (5'-UTR, CDS, or 3'-UTR) the P-site falls in is assigned.
3. Read count by nucleotide (metagene analysis)
	* [pending]
4. Read count by region (Figure 3A)
	* [pending]
5. Random forest analyses
	* [pending]
6. Comparative analyses
	* [pending]

